### PR TITLE
Destinations snowflake+bigquery: Fix string-json auto conversion for BigQuery

### PR DESCRIPTION
**Pull request by gisripa**
Summary
=======
* Add `BaseSqlGeneratorIntegrationTest#jsonStringifyTypes` to verify BigQuery and Snowflake behavior when it encounters json value for a string type specified in schema. 
* Fix BigQuery SQL to allow inserting non string types as a stringified value in the column instead of SQL null
* Minor nit to clean up if-else branching to just if-return and default return in `BigQuerySqlGenerator.java`

This addresses https://github.com/airbytehq/airbyte/issues/29524

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  destination-bigquery test report (commit [`4546fdeee4`](https://github.com/airbytehq/airbyte/pull/31358/commits/4546fdeee4cba2beed8eff9535d4e43d9350799e)) - ❌

⏲️  Total pipeline duration: 07mn59s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Build connector tar                                                  | ✅        |
| Build destination-bigquery docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                            | ✅        |
| Java Connector Integration Tests                                     | ✅        |
| Validate metadata for destination-bigquery                           | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ❌        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697141586/4546fdeee4cba2beed8eff9535d4e43d9350799e/destination-bigquery/2.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:4546fdeee4cba2beed8eff9535d4e43d9350799e)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java [hunk 1]
```diff
@@ -121,7 +121,9 @@ private String extractAndCast(final ColumnId column, final AirbyteType airbyteTy
       // This is guaranteed to not be a Union, so we won't recurse infinitely
       final AirbyteType chosenType = u.chooseType();
       return extractAndCast(column, chosenType, forceSafeCast);
-    } else if (airbyteType instanceof Struct) {
+    }
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java:124]: is this change necessary? I personally like the if/else chain (b/c then I can't accidentally put something important between two branches)

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java [hunk 2]
```diff
@@ -148,24 +152,40 @@ OR JSON_TYPE(PARSE_JSON(JSON_QUERY(`_airbyte_data`, '$."${column_name}"'), wide_
...
       // airbyte_data to json
       // and json_query it directly (which preserves nulls correctly).
       return new StringSubstitutor(Map.of("column_name", escapeColumnNameForJsonPath(column.originalName()))).replace(
           """
           JSON_QUERY(PARSE_JSON(`_airbyte_data`, wide_number_mode=>'round'), '$."${column_name}"')
           """);
+    }
+
+    if (airbyteType == AirbyteProtocolType.STRING) {
+      // Special case String to only use json value for type string and parse the json for others
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java:168]:
```suggestion
      // Special case String to only use json value for type string and parse the json for others
      // Naive json_value returns NULL for object/array values
```

**Review by edgao — APPROVED** [review]
nice! Super clean test + SQL implementation, and this is definitely solving a real problem for our users. alex C is going to be happy :)

(I had two small comments but overall :shipit:)

All that's left is to go through the publish process :rocket: (let's chat about this later)

**Comment by airbyte-oss-build-runner** [comment]
##  destination-bigquery test report (commit [`85c74c8ce1`](https://github.com/airbytehq/airbyte/pull/31358/commits/85c74c8ce18f65a53cf56f7f1ee3c3970b3a0c62)) - ❌

⏲️  Total pipeline duration: 08mn04s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Build connector tar                                                  | ✅        |
| Build destination-bigquery docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                            | ✅        |
| Java Connector Integration Tests                                     | ✅        |
| Validate metadata for destination-bigquery                           | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ❌        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697148585/85c74c8ce18f65a53cf56f7f1ee3c3970b3a0c62/destination-bigquery/2.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:85c74c8ce18f65a53cf56f7f1ee3c3970b3a0c62)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`85c74c8ce1`](https://github.com/airbytehq/airbyte/pull/31358/commits/85c74c8ce18f65a53cf56f7f1ee3c3970b3a0c62)) - ✅

⏲️  Total pipeline duration: 08mn47s 

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Build connector tar                                                   | ✅        |
| Build destination-snowflake docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                             | ✅        |
| Java Connector Integration Tests                                      | ✅        |
| Validate metadata for destination-snowflake                           | ✅        |
| Connector version semver check                                        | ✅        |
| QA checks                                                             | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697148585/85c74c8ce18f65a53cf56f7f1ee3c3970b3a0c62/destination-snowflake/3.2.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:85c74c8ce18f65a53cf56f7f1ee3c3970b3a0c62)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`47a162520d`](https://github.com/airbytehq/airbyte/pull/31358/commits/47a162520d08fdf69a8bdce4e42117e3325fb381)) - ✅

⏲️  Total pipeline duration: 10mn06s 

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Build connector tar                                                   | ✅        |
| Build destination-snowflake docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                             | ✅        |
| Java Connector Integration Tests                                      | ✅        |
| Validate metadata for destination-snowflake                           | ✅        |
| Connector version semver check                                        | ✅        |
| QA checks                                                             | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697149734/47a162520d08fdf69a8bdce4e42117e3325fb381/destination-snowflake/3.2.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:47a162520d08fdf69a8bdce4e42117e3325fb381)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-bigquery test report (commit [`47a162520d`](https://github.com/airbytehq/airbyte/pull/31358/commits/47a162520d08fdf69a8bdce4e42117e3325fb381)) - ❌

⏲️  Total pipeline duration: 06mn50s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Build connector tar                                                  | ✅        |
| Build destination-bigquery docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                            | ✅        |
| Java Connector Integration Tests                                     | ✅        |
| Validate metadata for destination-bigquery                           | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ❌        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697149734/47a162520d08fdf69a8bdce4e42117e3325fb381/destination-bigquery/2.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:47a162520d08fdf69a8bdce4e42117e3325fb381)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`462e092987`](https://github.com/airbytehq/airbyte/pull/31358/commits/462e0929870203b883304b64a955ba173b7437ff)) - ✅

⏲️  Total pipeline duration: 04mn10s 

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Build connector tar                                                   | ✅        |
| Build destination-snowflake docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                             | ✅        |
| Java Connector Integration Tests                                      | ✅        |
| Validate metadata for destination-snowflake                           | ✅        |
| Connector version semver check                                        | ✅        |
| QA checks                                                             | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697151050/462e0929870203b883304b64a955ba173b7437ff/destination-snowflake/3.2.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:462e0929870203b883304b64a955ba173b7437ff)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-bigquery test report (commit [`462e092987`](https://github.com/airbytehq/airbyte/pull/31358/commits/462e0929870203b883304b64a955ba173b7437ff)) - ✅

⏲️  Total pipeline duration: 07mn02s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Build connector tar                                                  | ✅        |
| Build destination-bigquery docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                            | ✅        |
| Java Connector Integration Tests                                     | ✅        |
| Validate metadata for destination-bigquery                           | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697151050/462e0929870203b883304b64a955ba173b7437ff/destination-bigquery/2.1.3/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:462e0929870203b883304b64a955ba173b7437ff)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`540def4d9a`](https://github.com/airbytehq/airbyte/pull/31358/commits/540def4d9a78ef27155d7b6d47387be9c948daa5)) - ✅

⏲️  Total pipeline duration: 04mn15s 

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Build connector tar                                                   | ✅        |
| Build destination-snowflake docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                             | ✅        |
| Java Connector Integration Tests                                      | ✅        |
| Validate metadata for destination-snowflake                           | ✅        |
| Connector version semver check                                        | ✅        |
| QA checks                                                             | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697152663/540def4d9a78ef27155d7b6d47387be9c948daa5/destination-snowflake/3.2.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:540def4d9a78ef27155d7b6d47387be9c948daa5)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-bigquery test report (commit [`540def4d9a`](https://github.com/airbytehq/airbyte/pull/31358/commits/540def4d9a78ef27155d7b6d47387be9c948daa5)) - ✅

⏲️  Total pipeline duration: 07mn56s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Build connector tar                                                  | ✅        |
| Build destination-bigquery docker image for platform(s) linux/x86_64 | ✅        |
| Java Connector Unit Tests                                            | ✅        |
| Java Connector Integration Tests                                     | ✅        |
| Validate metadata for destination-bigquery                           | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/gireesh_29524/1697152663/540def4d9a78ef27155d7b6d47387be9c948daa5/destination-bigquery/2.1.3/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:540def4d9a78ef27155d7b6d47387be9c948daa5)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Pull request merged**